storage engine: Generic typed key-value pairs

The type of keys and values that are used in applications may vary, so use generics.
**Workspace branch** for this issue: [kvdb~generic-store](https://github.com/de-sh/kvdb/tree/generic-store).

There appears to be a type conflict/error on [src/store.rs#L50](https://github.com/de-sh/kvdb/blob/generic-store/src/store.rs#L50)

typechk says:
```bash
error[E0308]: mismatched types
  --> src/store.rs:50:35
   |
21 | impl<A: Hash + Eq + Display, B: Display> Store<A,B> {
   |                              - this type parameter
...
50 |             Some(s) => Ok(B::from(s)),
   |                                   ^ expected type parameter `B`, found `&B`
   |
   = note: expected type parameter `B`
                   found reference `&B`
```
Couldn't figure out how to solve this.

--- a/src/repl.rs
+++ b/src/repl.rs
@@ -75,7 +75,7 @@ impl REPL {
                 },
                 StatementType::Rem => self.store.rem(key),
                 StatementType::Unk => {
-                    println!("db: command not found: {}", self.cmd);
+                    eprintln!("db: command not found: {}", self.cmd);
                     ExecResult::Failed
                 },
                 StatementType::Fail => ExecResult::Failed,

--- a/src/repl.rs
+++ b/src/repl.rs
@@ -8,8 +8,8 @@ use crate::store::{ExecResult, Store};
 pub struct REPL {
     /// User input read from the CLI, in string form.
     cmd: String,
-    /// Storage Engine currently in use.
-    store: Store,
+    /// Storage Engine used by the REPL with string based storage.
+    store: Store<String, String>,
 }
 
 impl REPL {

--- a/src/store.rs
+++ b/src/store.rs
@@ -1,4 +1,7 @@
 use std::collections::HashMap;
+use std::fmt::Display;
+use std::hash::Hash;
+
 
 /// Depicts whether an operation was successfully executed or not.
 pub enum ExecResult {
@@ -7,12 +10,15 @@ pub enum ExecResult {
 }
 
 /// The Storage Engine
-pub struct Store {
+pub struct Store<A,B> {
     /// A KV store in the form of in-memory HashMap.
-    storage: HashMap<String, String>,
+    /// Types A and B can be defined by the use case.
+    storage: HashMap<A, B>,
 }
 
-impl Store {
+/// As is clear from the implementation, types A and B must implement Display 
+/// to be 'printable'. While A must also implement Hash and Eq traits
+impl<A: Hash + Eq + Display, B: Display> Store<A,B> {
     /// Creates a new Storage Engine.
     pub fn new() -> Self {
         Self {
@@ -21,7 +27,7 @@ impl Store {
     }
 
     /// Operates HashMap::insert()
-    pub fn set(&mut self, key: String, value: String) -> ExecResult {
+    pub fn set(&mut self, key: A, value: B) -> ExecResult {
         // Fails if key already points to another value, else stores key-value pair and returns success.
         if self.storage.contains_key(&key) {
             eprintln!(
@@ -37,17 +43,17 @@ impl Store {
     }
 
     /// Operates HashMap::get() and fails if key-value pair doesn't
-    /// exist, else returns (string, success).
-    pub fn get(&self, key: String) -> Result<String, ExecResult> {
+    /// exist, else returns (value, success).
+    pub fn get(&self, key: A) -> Result<B, ExecResult> {
         match self.storage.get(&key) {
             None => Err(ExecResult::Failed),
-            Some(s) => Ok(s.to_string()),
+            Some(s) => Ok(B::from(s)),
         }
     }
 
     /// Operates HashMap::remove() and fails if the key-value pair
     /// doesn't exist, else removes it and returns success.
-    pub fn rem(&mut self, key: String) -> ExecResult {
+    pub fn rem(&mut self, key: A) -> ExecResult {
         match self.storage.remove(&key) {
             Some(val) => {
                 println!("Removed: {} -> {}", key, val);

--- a/src/store.rs
+++ b/src/store.rs
@@ -18,7 +18,7 @@ pub struct Store<A,B> {
 
 /// As is clear from the implementation, types A and B must implement Display 
 /// to be 'printable'. While A must also implement Hash and Eq traits
-impl<A: Hash + Eq + Display, B: Display> Store<A,B> {
+impl<A: Hash + Eq + Display, B: Display + Clone> Store<A,B> {
     /// Creates a new Storage Engine.
     pub fn new() -> Self {
         Self {
@@ -43,11 +43,11 @@ impl<A: Hash + Eq + Display, B: Display> Store<A,B> {
     }
 
     /// Operates HashMap::get() and fails if key-value pair doesn't
-    /// exist, else returns (value, success).
+    /// exist, else returns value on success.
     pub fn get(&self, key: A) -> Result<B, ExecResult> {
         match self.storage.get(&key) {
             None => Err(ExecResult::Failed),
-            Some(s) => Ok(B::from(s)),
+            Some(s) => Ok(B::from(s.clone())),
         }
     }
 

